feat: break up NAJ dependencies into packages

andy-haynes opened the pull request:
## Motivation
NAJ components have become more tightly coupled over time, leading to more complexity for new features and making it difficult for new contributors. By splitting up the NAJ codebase into packages, we can start to alleviate some of these issues and address longstanding community issues with NAJ runtime environments.

This PR is the first step towards a more modular NAJ but is largely a refactor of existing code. The new packages will be referenced and exported from NAJ, but the modules and their tests will be encapsulated by the new packages.

## Description
![image](https://user-images.githubusercontent.com/36863574/212437291-82b3764c-bd2c-43c2-bb38-9669bd775b14.png)



## Checklist
- [ ] Read the [contributing](https://github.com/near/near-api-js/blob/master/CONTRIBUTING.md) guidelines
- [ ] Commit messages follow the [conventional commits](https://www.conventionalcommits.org/) spec
- [ ] Performed a self-review of the PR
- [ ] Added automated tests
- [ ] Manually tested the change

andy-haynes commented on packages/crypto/src/key_pair_ed25519.ts:
```diff
@@ -0,0 +1,59 @@
+ * This class provides key pair functionality for Ed25519 curve:
+ * generating key pairs, encoding key pairs, signing and verifying.
+ */
+export class KeyPairEd25519 extends KeyPairBase {
```
By extending `KeyPairBase` rather than `KeyPair`, this becomes a breaking change by no longer including the static `fromString` method now on the abstract `KeyPair` class (which is now a sibling to `KeyPairEd25519`). Use of this now-deprecated method in practice would be circuitous and confusing, but its removal is still a breaking change.

This change isn't strictly necessary but is the only way to break the circular dependency across `PublicKey`, `KeyPair`, and `KeyPairEd25519`. This restructuring enables a more substantial decoupling later on, which involves moving `PublicKey` to `@near-js/types` and consolidating all `ed25519` library usage in this module.

frol reviewed (approved):
@andy-haynes This is a massive effort! I scanned through all the changes and left a couple of comments. I also run the tests locally and they pass. I believe this is in a pretty good shape, and I would merge it and iterate from there.

P.S. I feel it is a good time to revisit the 4-space indentation in JS/TS/JSON files. I would go with the 2-space style which is more common in JS world AFAIK.

andy-haynes commented on packages/providers/src/index.ts:
```diff
@@ -0,0 +1,4 @@
+export { exponentialBackoff } from './exponential-backoff';
+export { JsonRpcProvider } from './json-rpc-provider';
+export { Provider } from './provider';
+export { fetchJson } from './fetch_json';
```
I refrained from renaming files to minimize changes. I typically use snake case so that was what I used for new files in the absence of convention. Definitely agree that this should be cleaned up and enforced to some extent.

andy-haynes commented on packages/accounts/test/account.access_key.test.js:
```diff
@@ -0,0 +1,97 @@
+const { KeyPair } = require('@near-js/crypto');
```
I have a PR in progress to convert these to ESM but nothing in the works to support TypeScript tests, though I'm not opposed.

andy-haynes commented on packages/accounts/src/account_2fa.ts:
```diff
@@ -0,0 +1,281 @@
+'use strict';
```
I don't believe so since we compile with `alwaysStrict` enabled. This module is new but was copied from `account_multisig.ts` which includes it.

andy-haynes commented:
@frol thank you for the review! I've rebased the package lock to incorporate the new package and addressed some of the low-hanging fruit from your feedback. One additional change was necessary as a workaround to get `/lib/transactions.d.ts` to be emitted correctly (see comment) but otherwise everything looks good.

> P.S. I feel it is a good time to revisit the 4-space indentation in JS/TS/JSON files. I would go with the 2-space style which is more common in JS world AFAIK.

This is my preference as well, I would like to follow this work up with some enhanced tsconfig and eslint configuration.

> P.S.S. Are there known issues to be addressed in this PR?

Not directly, though usage of `@near-js` packages in place of `near-api-js` may alleviate some import issues.

frol reviewed (approved):
Looks good to me! Let's merge it?
